hy/compiler.py

```diff
@@ -598,6 +621,166 @@ def compile_raise_expression(self, expr, root, exc, cause):
             expr, type=ret.expr, exc=exc,
             inst=None, tback=None, cause=cause)
 
+    notsym = lambda *dissallowed: some(
+        lambda x: isinstance(x, Symbol) and str(x) not in dissallowed
+    )
+    keepsym = lambda wanted: some(lambda x: isinstance(x, Symbol) and x == wanted)
+    _match_value = SYM | LITERAL
+    _pattern = forward_decl()
+    _assignment = maybe(sym(":as") + SYM)
+    _pattern.define(
+        _match_value + _assignment
+        | brackets(many(_pattern | unpack("iterable"))) + _assignment
+        | pexpr(keepsym("|"), many(_pattern)) + _assignment
+        | pexpr(keepsym(","), many(_pattern)) + _assignment
+        | curly_brackets(many(LITERAL + _pattern), maybe(pvalue("unpack-mapping", SYM))) + _assignment
+        | pexpr(
+            notsym("|", ",", "unpack-mapping", "unpack-iterable"),
+            many(_pattern),
+            many(KEYWORD + _pattern),
+        ) + _assignment
```

Comment:
The ` + _assignment` can be factored out if you stick the alternations inside a set of parentheses, no?

Reply:
good catch! done